lib/salus/scanners/npm_audit.rb

```diff
@@ -1,68 +1,287 @@
+      red: 31,
+      yellow: 33,
+      green: 32
+    }.freeze
+
+    AUDIT_COMMAND = 'npm audit --json'.freeze
+    LOCKFILE_COMMAND = 'npm install --package-lock-only'.freeze
 
     def run
-      exceptions = @config['exceptions']
+      sanity_check_exceptions
 
-      created_package_lock = false
-      if !@repository.package_lock_json_present?
-        Dir.chdir(@repository.path_to_repo) do
-          report_info(
-            'package_lock_missing',
-            'No package.lock file was found, so we generated one for you. As long as '\
-              'you use yarn, this should be fine. If not, please check in a package-lock.json '\
-              'file into your source control.'
+      @exceptions =
+        @config.fetch('exceptions', [])
+          .map { |ex| ex['advisory_id']&.to_s }
+          .compact
+
+      @stream = StringIO.new
+
+      okay = run_audit!
+      report_stdout(@stream.string)
+
+      if okay
+        report_success
+      else
+        report_failure
+      end
+    end
+
+    def should_run?
+      @repository.package_json_present? ||
+        @repository.package_lock_json_present? ||
+        @repository.yarn_lock_present?
+    end
+
+    def sanity_check_exceptions
+      @config['exceptions']&.each do |ex|
+        if !ex.is_a?(Hash) || ex.keys.sort != %w[advisory_id changed_by notes]
+          report_error(
+            'message' =>
+              "malformed exception: #{ex.inspect}; " \
+              'expected a hash with keys `advisory_id`, `changed_by`, `notes`'
           )
-          run_shell('npm i --package-lock-only')
-          created_package_lock = true
+          next
         end
       end
+    end
 
+    def run_audit!
       Dir.chdir(@repository.path_to_repo) do
-        shell_return = run_shell('npm audit --json')
-
-        if shell_return[:exit_status].success?
-          report_success
-        else
-          # Parse output
-          npm_audit_report = JSON.parse(shell_return[:stdout])
-          report_stdout(shell_return[:stdout])
-
-          # Report scan output
-          report_info('npm_audit_output', npm_audit_report)
-
-          # Report ignored advisories after validating the keys all exist
-          exceptions&.each do |exception|
-            if exception.keys.sort != %w[advisory_id changed_by notes]
-              report_error("The exception #{exception} doesn't have a proper format! Please ensure"\
-              ' that each exception has an `advisory_id`, `changed_by`, and `notes` field!')
+        ensure_package_lock! do
+          raw = run_shell(AUDIT_COMMAND).fetch(:stdout)
+          json = JSON.parse(raw, symbolize_names: true)
+
+          if json.key?(:error)
+            code = json[:error][:code] || '<none>'
+            summary = json[:error][:summary] || '<none>'
+
+            message =
+              "`#{AUDIT_COMMAND}` failed unexpectedly (error code #{code}):\n" \
+              "```\n#{summary}\n```"
+
+            log(colorize(message, :red))
```

Comment:
use [`report_error()`](https://github.com/coinbase/salus/blob/master/lib/salus/scanners/base.rb#L64) here - this will ensure that the Salus report shows that this particular scanner threw an error instead of completing successfully.